sympy/solvers/solveset.py

```diff
@@ -1963,6 +2020,12 @@ def add_type(lhs, rhs, symbol, domain):
 
         if lhs.is_Add:
             result = add_type(lhs, rhs, symbol, domain)
+        elif _is_lambert(lhs - rhs):
+            # try to get solutions in form of Lambert since `a*f(x) = c`
+            # will not come under lhs.is_add and which can't be solved by
+            # invert_complex for lambert type. (it's exception because of many
+            # forms and code of such cases are covered in `_solve_lambert`).
+            result = _solve_as_lambert(lhs, rhs, symbol, domain)
```

Comment:
It would be better if we could handle this as part of `invert_complex` or `invert_real` I think. This is because `invert_complex` is used in other places as well so improving that function improves many parts of `solveset`.

Reply:
> Maybe there's something I don't understand but is it not straight-forward to move the _is_lambert check into invert_complex?

Yep, that's what I've thought also about it just now : ) !
I think I should keep a check for _is_lambert in the `invert_complex` function and from there divert the logic towards `_solve_lambert` function !

Reply:
I think it's better to improve lambert for `invert` after improving `_is_lambert` (https://github.com/sympy/sympy/pull/21025#issue-583569514), to avoid a collision as there is a lot of dependency of different functions with `invert`.

Reply:
Do you mean that adding this to `invert` leads to lots of test failures?

If so then that's good because we can see how the lambert code works in many examples and fix any issues that come up.

Reply:
> Do you mean that adding this to invert leads to lots of test failures?
> 
> If so then that's good because we can see how the lambert code works in many examples and fix any issues that come up.

No, that's not the point I want to convey instead you may see here https://github.com/sympy/sympy/blob/741dd194e2a825104631e4fa4605b0b060fc40c1/sympy/solvers/solveset.py#L1761-L1768

It will try to consider all `(Pow, exp, log)` types as `lambert`. So that's the reason to improve `_is_lambert` function (i.e. to make it more specific)
The reason I've not included that part in this PR is to make the difference less. ( The better version of _is_lambert is https://github.com/sympy/sympy/pull/20824#discussion_r584329684)
https://github.com/sympy/sympy/blob/a8e79f2e1a95a3985c5fa0071a6e421053d9b273/sympy/solvers/solveset.py#L1817-L1831

(Here, the documentation has to be improved.)

Reply:
Well it looks like the `_is_lambert` function needs to be improved. I don't understand why it works in one place but not in the other though.

Reply:
The lines of code from `def _invert(f_x, y, x, domain=S.Complexes):`
https://github.com/sympy/sympy/blob/741dd194e2a825104631e4fa4605b0b060fc40c1/sympy/solvers/solveset.py#L171-L185
```
if _is_lambert(f_x ):
     return x, _solve_as_lambert(x1, s.args[0], x, domain)
```
If we append this code at L-175 then there will be no need for `_is_lambert` declaration in `_transolve` because of this line L-1019 below (in _solveset) https://github.com/sympy/sympy/blob/741dd194e2a825104631e4fa4605b0b060fc40c1/sympy/solvers/solveset.py#L994-L1020

Because it will solve all types of lambert from here itself.
Should I proceed with this step ?
Please give your opinion !

Reply:
> Should I proceed with this step?
> Please give your opinion!

Kindly, reply to this detailed comment.

Reply:
I think it should be treated the same as `_is_modular` ( which is in `L-1064` in `_solveset`  function) .

Reply:
> Kindly, reply to this detailed comment.

I don't understand exactly what you are proposing. The way that `is_modular` is handled doesn't look very good to me. I don't see why it isn't just part of `invert_real`.

Reply:
> I don't understand exactly what you are proposing. The way that `is_modular` is handled doesn't look very good to me. I don't see why it isn't just part of invert_real.

I think it's because invert_real works with a lot of recursions whereas `is_lambert` and `_is_modular` don't go through that process instead they directly give solutions in one go...

Reply:
> I don't understand exactly what you are proposing.

The latest proposal is to add `elif _is_lambert(.. ` at `L-1066` at `_solveset` function, instead of `_transolve` or `invert`.

Reply:
Would it not be useful to have recursions with modular and Lambert equations as well though?

For example:
```python
In [1]: f = lambda x: x*exp(x)                                                                                                    

In [2]: f(f(x))-y                                                                                                                 
Out[2]: 
         x    
   x  x⋅ℯ     
x⋅ℯ ⋅ℯ     - y

In [3]: solve(_, x)                                                                                                               
Out[3]: 
⎡ ⎛   -W(y)⎞⎤
⎣W⎝y⋅ℯ     ⎠⎦
```

Reply:
I think for modular there is already `_invert_modular` but it's not integrated with the `invert` function.

Also, I think in a similar way `invert_lambert` should also be developed and then it should be integrated with `invert` (along with `_invert_modular`) so that `invert` could handle both.

Reply:
My 2nd proposal is,

> If we add this code at L-175 then there will be no need for _is_lambert declaration in _transolve because of this line L-1019 below (in _solveset) (https://github.com/sympy/sympy/pull/21025#discussion_r591116862)

Adding code in `invert` then leads to removal of code from `_transolve` (since the solution will be returned from `L-1019` itself.)

Reply:
Do you think the 2 nd proposal is a better one instead of just adding lines of code in `_solveset`?

Reply:
I don't understand exactly what you are proposing. I think it's easiest if you just push your suggestion to the PR if you want me to comment on it.

In general though I think that it is better to add these things in the invert code.

Reply:
> In general though I think that it is better to add these things in the invert code.

In this case, we do not need to have code in `_transolve` !

Reply:
I just gave my opinion!

Please if you want to ask a question you need to actually explain what the question is. Do not presume that I will go and reread all of the messages on Github from all of your PRs and understand what you mean. As I already said it is much easier for me to understand what you mean if you push the changes to a PR.